src/format.rs

```diff
@@ -0,0 +1,45 @@
+//! Macros for string formatting.
+
+/// Conditionally perform string formatting.
+///
+/// If `$enabled` is true, then do the formatting and return a `Cow::Owned`.
+///
+/// Otherwise, just return the borrowed (often `'static`) string
+/// `$borrowed`.
+///
+/// When `$enabled` is false, this avoids the overhead of allocating
+/// and writing to a buffer, as well as any overhead or side effects
+/// of the format arguments.
+///
+/// # Example
+///
+/// You can use `format_if` to implement a detailed error logging facility
+/// that can be enabled at runtime.
+///
+/// ```
+/// # #[macro_use] extern crate mac;
+/// # fn main() {
+/// let formatted = format_if!(true, "Vague error", "Error code {:?}", 3);
+///
+/// assert_eq!(&formatted[..], "Error code 3");
+/// assert!(formatted.is_owned());
+///
+/// let not_formatted = format_if!(false, "Vague error", "Error code {:?}", {
+///     // Note that the argument is not evaluated.
+///     panic!("oops");
+/// });
+///
+/// assert_eq!(&not_formatted[..], "Vague error");
+/// assert!(not_formatted.is_borrowed())
+/// # }
+/// ```
+#[macro_export]
+macro_rules! format_if {
```

Comment:
I look at this macro and I see a verbosity switch. The really cool thing here is being able to switch being verbose (full, formatted, allocated error) and quick (short, borrowed error).

In light of this, it might be neat to re-use the verbosity levels and setup from the log crate and branch on that. If the logging setup isn't ideal, we can also just use the verbosity levels and leave them unconnected. Thoughts?